### ISSUE TITLE
|^ crash when $ arm is present (or, why I learned to stop worrying and love unhygenic macros)

### ISSUE THREAD

**Issue by dhaffey** [urbit/urbit#616]
```
dojo> |^  a  ++  a  +($)  ++  $  42  --  ::  address 0x7fff525afff8 out of loom!
```

However,

```
dojo> =<  foo  |%  ++  foo  +($)  ++  $  42  --
43
```

so I don't _think_ this is a benign stack overflow, but I may be misunderstanding something.

**Comment by dhaffey** [comment]
Definitely seems like |^ needs a shower:

    dojo> ~(open ap (ream '|^  a  ++  a  +($)  ++  $  42  --'))
    [ %tsgr
      p=[%brcn p={[p=%$ q=[%ash p=[%cnzz p=~[%a]]]] [p=%a q=[%ash p=[%dtls p=[%cnzz p=~[%$]]]]]}]
      q=[%cnzy p=%$]
    ]

**Comment by juped** [comment]
I think this is an area prone to bitrot. @cgyarvin should weigh in.

**Comment by ohAitch** [comment]
This seems like a vanilla infinite loop, #354 perhaps being relevant to why it aborts instead of crashing properly.

As for why this causes an infinite loop, compare `=<  $  |%  foo  +($)  ++  $  42   ++  $  foo`. Which the parser/compiler also should be catching but isn't, yes.

**Comment by ohAitch** [comment]
(Something useful would be a `+-put` variant that crashes if the value already exists, and `+-gas` using that instead of actual `+-put`. Perhaps with old-`gas` accessible under an alternate name for any existing uses that *do* want to allow duplicates.)

**Comment by cgyarvin** [comment]
Yes, this is a vanilla infinite loop, since |^ puts the initial twig in as
a $ foot.   I'm not a big fan of |^, it does need a shower... and yes, it's
a bug that we don't catch duplicate arms.

**Comment by ohAitch** [comment]
Should `|^` just desugar to `=< |%`, or do we use the recursive capabilities somewhere?

**Comment by joemfb** [comment]
This is still a problem: duplicate arms are prevent, but only by the parser ... Should `|^` expand to `=<  |%`?

**Comment by juped** [comment]
Further discussion should go into the fix thread, urbit/arvo#343
